### ISSUE TITLE
`createJSONStorage`: localStorage is not defined

### ISSUE THREAD

**Issue by rtritto** [pmndrs/jotai#2581]
## Summary
```console
ReferenceError: localStorage is not defined

  1 | import { atomWithStorage, createJSONStorage } from 'jotai/utils'
  2 |
> 3 | const storage = createJSONStorage(() => localStorage)
    |                                         ^
  4 |
  5 | const darkModeState = atomWithStorage('darkMode', true, storage)
```

## Details
Next.js: `14.2.3`
jotai: `2.8.2`

It correctly works with jotai <= 2.8.1.

Issue probably generated by #2539

Similar issue #949

**Comment by fernandojbf** [comment]
Having the same issue for sessionStorage. 

I believe 2.8.2 start calling the storage fn on server side.

**Comment by dai-shi** [comment]
Ohhhh, that was it. https://github.com/pmndrs/jotai/pull/2539#discussion_r1589803732
Thanks for reporting.

#2539 broke it.
@mhsattarian Would you like to fix?

**Comment by mhsattarian** [comment]
@dai-shi on it.

**Comment by dai-shi** [comment]
Can any of you confirm if #2585 fixes your case?
https://ci.codesandbox.io/status/pmndrs/jotai/pr/2585
☝️ Find "Local Install Instructions"

**Comment by rtritto** [comment]
@dai-shi it works, thanks!

PS: I'm using yarn v4: this version should be added in the second step of "Local Install Instructions" of CodeSandbox CI

**Comment by dai-shi** [comment]
Thanks for confirming!

(Yeah, but it's out of my control.)

**Comment by omridevk** [comment]
@dai-shi 
Checking with version 2.8.4
I am still getting:
```
15:18:41 [vite] Error when evaluating SSR module /src/components/wizard/wizard.tsx:
|- ReferenceError: sessionStorage is not defined
```

my code:

`const storage = createJSONStorage<WizardAtom>(() => sessionStorage)`

**Comment by omridevk** [comment]
it is hard for me create a reproduction as I am using vite + fastify to create an SSR output

**Comment by dai-shi** [comment]
Does it work with v2.8.1? Otherwise, it's probably a different thing.

**Comment by omridevk** [comment]
Nope, same issue.

**Comment by omridevk** [comment]
```
15:52:21 [vite] Error when evaluating SSR module /src/atoms/wizard-data-atom.ts:
|- ReferenceError: localStorage is not defined
    at eval (/src/atoms/wizard-data-atom.ts:7:63)
    at Object.getItem (/node_modules/jotai/esm/vanilla/utils.mjs:372:31)
    at Module.atomWithStorage (/node_modules/jotai/esm/vanilla/utils.mjs:419:25)
    at eval (/src/atoms/wizard-data-atom.ts:8:46)
    at async instantiateModule (/vite/dist/node/chunks/dep-C1-ZB6nQ.js:53451:5)
```

**Comment by dai-shi** [comment]
> It correctly works with jotai <= 2.8.1.

In #2581, we solved the issue specific in v2.8.2.
If v2.8.1 doesn't solve your case, please open a new bug report 👉 https://github.com/pmndrs/jotai/discussions/new?category=bug-report
Or, a question https://github.com/pmndrs/jotai/discussions/new?category=q-a if you don't have a reproduction.

**Comment by omridevk** [comment]
Thanks @dai-shi for the quick response.
https://github.com/pmndrs/jotai/discussions/2639
And I will open an issue once I can reproduce it in stackblitz/codesandbox or a repo. 
Really love this lib by the. way, appreciate all your hard work sensei
